### PR TITLE
feat(CDM): add a new resource: CDM link

### LINKED ISSUE #1775: [CDM] Add a new resource to manage CDM link
I'd like to add CDM link resource.

### Current Terraform Version

```
...
```

### PR DESCRIPTION
**Pull request by hmbig2**
**What this PR does / why we need it**:
add a new resource: CDM link

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1775 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/cdm' TESTARGS='-run=TestAccResourceCdmLink_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cdm -v -run=TestAccResourceCdmLink_basic -timeout 360m -parallel 4
=== RUN   TestAccResourceCdmLink_basic
=== PAUSE TestAccResourceCdmLink_basic
=== CONT  TestAccResourceCdmLink_basic
--- PASS: TestAccResourceCdmLink_basic (605.73s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cdm       605.863s
```

### CODE REVIEW

File: docs/resources/cdm_link.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+resource "huaweicloud_cdm_link" "obsLink" {
+  name       = var.obs_link_name
+  connector  = "obs-connector"
+  cluster_id = var.cdm_cluster_id
+
+  config = {
+    "storageType" = "OBS"
+    "server"      = trimprefix(huaweicloud_obs_bucket.bucket.bucket_domain_name, "${huaweicloud_obs_bucket.bucket.bucket}.")
```
Comment (chengxiangdong) [inline docs/resources/cdm_link.md:34]: can the function of removing the prefix be handled in the code?

File: docs/resources/cdm_link.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    "port"        = "443"
+  }
+  access_key   = var.access_key
+  security_key = var.security_key
+}
+```
+
+### link to MySql
```
Comment (chengxiangdong) [inline docs/resources/cdm_link.md]: s/link/Link

File: docs/resources/cdm_link.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  - **Link to MongoDB**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0072.html)
+  - **Link to Kafka**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0035.html)
+  - **Link to DIS**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0075.html)
+  - **Link to Elasticsearch/Cloud Search Service**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0076.html)
+  - **Link to DLI**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0070.html)
+  - **Link to CloudTable OpenTSDB**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0089.html)
+  - **Link to Amazon S3**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0094.html)
+  - **Link to DMS Kafka**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0095.html)
+
+-> Please remove the `linkconfig.` in the parameter key listed in the document. Configuration parameters such as
```
Comment (chengxiangdong) [inline docs/resources/cdm_link.md:119]: It is recommended to be consistent with other note styles, use `-> **NOTE:**`

File: docs/resources/cdm_link.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID in format of **cluster_id/link_name**. It is composed of the ID of CDM cluster and the name
+ of job, separated by a slash.
+
+## Import
+
+The link can be imported by `id`,It is composed of the ID of CDM cluster and the name of job, separated by a slash.
```
Comment (chengxiangdong) [inline docs/resources/cdm_link.md]: s/`id`,It/`id`, It

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+				Name:             d.Get("name").(string),
+				ConnectorName:    d.Get("connector").(string),
+				Enabled:          utils.Bool(d.Get("enabled").(bool)),
+				LinkConfigValues: buildLinkConfigParamter(d),
+			},
+		},
+	}
+
+	clusterId := d.Get("cluster_id").(string)
+	logp.Printf("[DEBUG] Creating CDM link opts: %#v", opts)
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go]: here, the sensitive fields such as password, ak, sk, etc. will be printed in the log, which are not compliant and not allowed.

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+	d.SetId(fmt.Sprintf("%s/%s", clusterId, rst.Name))
+
+	return resourceCdmLinkRead(ctx, d, meta)
+}
+
+func buildLinkConfigParamter(d *schema.ResourceData) link.LinkConfigs {
+	var configs []link.Input
+	configRaw := d.Get("config").(map[string]interface{})
+	connector := d.Get("connector").(string)
+	if len(configRaw) > 0 {
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go]: If the length of configRaw is 0, suggest return an error directly, and do not continue to call the API.

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+		for k, v := range configRaw {
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go]: is there a case where multiple configs share a set of authentication information?
If so, does the API support it?

File: docs/resources/cdm_link.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+ For example,
+
+```
+terraform import huaweicloud_cdm_link.test b11b407c-e604-4e8d-8bc4-92398320b847/linkName
```
Comment (chengxiangdong) [inline docs/resources/cdm_link.md:144]: some fields cannot be imported, and descriptions should be added, please refer to: [/docs/resources/dcs_instance.md](https://github.com/huaweicloud/terraform-provider-huaweicloud/blob/master/docs/resources/dcs_instance.md)

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+			{
+				Name:             newName,
+				ConnectorName:    d.Get("connector").(string),
+				Enabled:          utils.Bool(d.Get("enabled").(bool)),
+				LinkConfigValues: buildLinkConfigParamter(d),
+			},
+		},
+	}
+
+	logp.Printf("[DEBUG] Update CDM link opts: %#v", opts)
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go]: sensitive fields will be printed to the log, please block the processing

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+
+	_, updateErr := link.Update(client, clusterId, linkName, opts)
+	if updateErr != nil {
+		return fmtp.DiagErrorf("Error update CDM link: %s", updateErr)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", clusterId, newName))
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go:294]: is this redundant?

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func ParseLinkInfoFromId(id string) (clusterId, linkName string) {
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go:305]: it seems that this function will not be used outside the package,
it is recommended to use `parseLinkInfoFromId`

File: docs/resources/cdm_link.md (hunk 3 above)
Comment (hmbig2) [reply docs/resources/cdm_link.md:119]: this is standard style,  and `:` is redundant.  so keep it as it is

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go (hunk 10 above)
Comment (hmbig2) [reply huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go:294]: no, name can be change

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+	idArrays := strings.Split(id, "/")
+	clusterId = idArrays[0]
+	linkName = idArrays[1]
```
Comment (chengxiangdong) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go:333]: when importing the id entered by the user, the array may be out of bounds

File: docs/resources/cdm_link.md (hunk 1 above)
Comment (hmbig2) [reply docs/resources/cdm_link.md:34]: no, According to different data sources, the config is different.  detail is listed in document

File: huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+	if err != nil {
+		return diag.Errorf("error creating CDM v1 client, err=%s", err)
+	}
+
+	linkConfigValues, err := buildLinkConfigParamter(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	opts := link.LinkCreateOpts{
```
Comment (ShiChangkuo) [inline huaweicloud/services/cdm/resource_huaweicloud_cdm_link.go:103]: suggest to add a validate before create the resource

File: docs/resources/cdm_link.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+  - **Link to MongoDB**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0072.html)
+  - **Link to Kafka**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0035.html)
+  - **Link to DIS**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0075.html)
+  - **Link to Elasticsearch/Cloud Search Service**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0076.html)
+  - **Link to DLI**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0070.html)
+  - **Link to CloudTable OpenTSDB**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0089.html)
+  - **Link to Amazon S3**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0094.html)
+  - **Link to DMS Kafka**: [configuration detail](https://support.huaweicloud.com/intl/en-us/api-cdm/cdm_02_0095.html)
+
+-> Please remove the `linkconfig.` in the parameter key listed in the document. Configuration parameters such as
```
Comment (ShiChangkuo) [inline docs/resources/cdm_link.md:119]: suggest to support a key both with prefix `linkconfig` or not

Comment (hmbig2) [reply docs/resources/cdm_link.md:119]: `.` the dot Will invalidate the get method.  so the user better ignores the character `linkconfig.`

File: docs/resources/cdm_link.md (hunk 3 above)
Comment (chengxiangdong) [reply docs/resources/cdm_link.md:119]: your writing style is different from the exists, we always use: `-> **NOTE:**`.
There are also different styles on the page display, which shouldn't be.

**Pull request merged**